src/tactic/move_add.lean

```diff
@@ -0,0 +1,358 @@
+  -- the current `do` block takes place where the reordered expression is not equal to the original
+  neq ← mk_app `eq [thyp, reordered],
+  pre ← pp reordered,
+  (_, prf) ← solve_aux neq $
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done }] <|>
+    `[{ simp only [mul_comm, mul_assoc, mul_left_comm], done }] <|>
+    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
+    "the initial goal equals:\n\n{pre}\n"),
+  if na.is_none then refine ``(eq.mpr %%prf _) else replace_hyp hyploc reordered prf >> skip,
+  return (ff, is_unused)
+
+section parsing_arguments_for_move_op
+setup_tactic_parser
+
+/-- `move_op_arg` is a single elementary argument that `move_op` takes for the
+variables to be moved.  It is either a `pexpr`, or a `pexpr` preceded by a `←`. -/
+meta def move_op_arg (prec : nat) : parser (bool × pexpr) :=
+prod.mk <$> (option.is_some <$> (tk "<-")?) <*> parser.pexpr prec
+
+/-- `move_pexpr_list_or_texpr` is either a list of `move_op_arg`, possibly empty, or a single
+`move_op_arg`. -/
+meta def move_pexpr_list_or_texpr : parser (list (bool × pexpr)) :=
+list_of (move_op_arg 0) <|> list.ret <$> move_op_arg tac_rbp <|> return []
+
+end parsing_arguments_for_move_op
+
+end move_op
+
+setup_tactic_parser
+open move_op
+
+/--  `move_op args locat op` is the non-interactive version of the main tactics `move_add` and
+`move_mul` of this file.  Given as input `args` (a list of terms of a sequence of operands),
+`locat` (hypotheses or goal where the tactic should act) and `op` (the operation to use),
+`move_op` attempts to perform the rearrangement of the terms determined by `args`.
+
+Currently, the tactic uses only `add/mul_comm, add/mul_assoc, add/mul_left_comm`, so other
+operations will not actually work.
+-/
+meta def move_op (args : parse move_pexpr_list_or_texpr) (locat : parse location) (op : pexpr) :
+  tactic unit := do
+locas ← locat.get_locals,
+tg ← target,
+let locas_with_tg := if locat.include_goal then locas ++ [tg] else locas,
+ner ← locas_with_tg.mmap (λ e, reorder_hyp op args e.local_pp_name <|> reorder_hyp op args none),
+let (unch_tgts, unus_vars) := ner.unzip,
+let str_unva := match
+  (return_unused args (unus_vars.transpose.map list.band)).map (λ e : bool × pexpr, e.2) with
+  | []   := ""
+  | [pe] := "'" ++ to_string pe ++"' is an unused variable"
+  | pes  := "'" ++ to_string pes ++"' are unused variables"
+  end,
+let str_tgts := match locat with
+  | loc.wildcard := if unch_tgts.band then "nothing changed" else ""
+  | loc.ns names := let linames := return_unused locas unch_tgts in
+      (if none ∈ return_unused names unch_tgts then "Goal did not change\n" else "") ++
+      (if linames ≠ [] then ("'" ++ to_string linames.reverse ++ "' did not change") else "")
+  end,
+guard (str_tgts ++ str_unva = "") <|>
+  fail (if str_tgts.length = 0 then str_unva else str_tgts ++ "\n" ++ str_unva),
+assumption <|> try (tactic.reflexivity reducible)
```

Comment:
Using `format` instead of `to_string` means the info-view knows how to wrap longer error messages:

```suggestion
meta def move_op (args : parse move_pexpr_list_or_texpr) (locat : parse location) (op : pexpr) :
  tactic unit := do
locas ← locat.get_locals,
tg ← target,
let locas_with_tg := if locat.include_goal then locas ++ [tg] else locas,
ner ← locas_with_tg.mmap (λ e, reorder_hyp op args e.local_pp_name <|> reorder_hyp op args none),
let (unch_tgts, unus_vars) := ner.unzip,
let str_unva := match
  (return_unused args (unus_vars.transpose.map list.band)).map (λ e : bool × pexpr, e.2) with
  | []   := []
  | [pe] := [format!"'{pe}' is an unused variable"]
  | pes  := [format!"'{pes}' are unused variables"]
  end,
let str_tgts := match locat with
  | loc.wildcard := if unch_tgts.band then [format!"nothing changed"] else []
  | loc.ns names := let linames := return_unused locas unch_tgts in
      (if none ∈ return_unused names unch_tgts
        then [format!"Goal did not change"] else []) ++
      (if linames ≠ [] then [format!"'{linames.reverse}' did not change"] else [])
  end,
[] ← pure (str_tgts ++ str_unva) | fail (format.intercalate "\n" (str_tgts ++ str_unva)),
assumption <|> try (tactic.reflexivity reducible)
```
You'll need to remove some trailing `\n`s from the test too.